chore: update contribution guideline by adding code style requirement (#592)
jina-ai/docarray: CONTRIBUTING.md

Run `git config blame.ignoreRevsFile .github/.git-blame-ignore-revs`

## Code style conventions:

Most of our codebase is written in Python. 

### Python version
DocArray is compatible with Python 3.7 and above, therefore we can't accept contribution that used features from the newest Python versions without ensuring compatibility with python 3.7

### Code formatting

> All of our Python codebase follow formatting standard. We are following the [PEP8](https://peps.python.org/pep-0008/) standard, and we require that every code contribution is formatted using [black](https://github.com/psf/black) with the default configurations.

Reviewer: ```suggestion
All of our Python codebase follows formatting standard. We are following the [PEP8](https://peps.python.org/pep-0008/) standard, and we require that every code contribution is formatted using [black](https://github.com/psf/black) with the default configurations.
```

> If you have installed the [pre-commit hooks](https://github.com/jina-ai/docarray/blob/main/CONTRIBUTING.md#install-pre-commit-hooks) the formatting should be automatic on every commit. Moreover, our CI will block contribution that does not respect this conventions.

Reviewer: ```suggestion
If you have installed the [pre-commit hooks](https://github.com/jina-ai/docarray/blob/main/CONTRIBUTING.md#install-pre-commit-hooks) the formatting should be automatic on every commit. Moreover, our CI will block contributions that does not respect this conventions.
```

### Type Hint

> Python is not a strongly typed programming language nevertheless the use of [type hints](https://docs.python.org/3/library/typing.html)  

Reviewer: ```suggestion
Python is not a strongly typed programming language, nevertheless the use of [type hints](https://docs.python.org/3/library/typing.html)  
```

contribute to a better codebase especially when reading, reviewing and refactoring. Therefore, we **highly** encourage every contribution

> to fully utilise type hints. In some particular case type hints can be cumbersome, therefore using type hints is not a hard requirement for contribution.

Reviewer: ```suggestion
to fully utilise type hints. In some particular cases type hints can be cumbersome, therefore using type hints is not a hard requirement for contribution.
```

Run `git config blame.ignoreRevsFile .github/.git-blame-ignore-revs`

## Code style conventions:

Most of our codebase is written in Python. 

### Python version
We support all the Python version starting with and above 3.7. It means that we require in every contribution that the 

> code is compatible with python 3.7, and therefore we can't accept contribution that used features from the newest python versions.

Reviewer: Does it make sense to say

> we can't accept contribution that used features from the newest python versions without ensuring compatibility with python 3.7

Author: done

> code is compatible with python 3.7, and therefore we can't accept contribution that used features from the newest python versions.

Reviewer: ```suggestion
code is compatible with python 3.7, and therefore we can't accept contributions that use features from the newest python versions.
```

### Code formatting

> All of our Python codebase follow formatting standard. We require that every code contribution is formatted using [black](https://github.com/psf/black)

Reviewer: if we're refferring to pep8 standard let's mention it (because that's what black does)

Author: u are right

Author: Black follow pep8 but go above and do some custom things. I can add that we follow pep8 but we still need to say that we follow black conventions

> All of our Python codebase follow formatting standard. We require that every code contribution is formatted using [black](https://github.com/psf/black)

Reviewer: ```suggestion
All of our Python codebase follows [pep8](https://peps.python.org/pep-0008/) style standard. We require that every code contribution is formatted using [black](https://github.com/psf/black)
```

We are using the default configuration.

> If you have installed the `pre-commit hooks` (see above) the formatting should be automatic on every commit. Moreover, our CI will block contribution that does not respect this conventions.

Reviewer: can we link the pre-commit hook section ?

### Type Hint

> Python is not a strongly typed programming language nevertheless the use of [type hint](https://docs.python.org/3/library/typing.html)  

Reviewer: ```suggestion
Python is not a strongly typed programming language nevertheless the use of [type hints](https://docs.python.org/3/library/typing.html)  
```

contribute to a better codebase especially when reading, reviewing and refactoring. Therefore, we **highly** encourage every contribution

> to fully utilise type hint. In some particular case type hint can be cumbersome, therefore using type hint is not a hard requirement for contribution.

Reviewer: ```suggestion
to fully utilise type hints. In some particular case type hints can be cumbersome, therefore using type hints is not a hard requirement for contribution.
```

> Contributions are expected to use type hint, especially in function signature, unless there is an arguably good reason not to do it.

Reviewer: ```suggestion
Contributions are expected to use type hints, especially in function signature, unless there is an arguably good reason not to do it.
```
